### ISSUE TITLE
[Feature Request] Use an URL instead of hostname to connect to a RPC server

### ISSUE THREAD

**Issue by NicolasDorier** [sparrowwallet/sparrow#774]
### Issue

As you can see here, while the field say `URL`, an URL doesn't work.

![image](https://user-images.githubusercontent.com/3020646/207211884-48e5b6d0-bbf0-4571-915e-8341dd4a1d70.png)

### Why should an URL be supported

I think an URL should be supported. I am planning to create some middleware to allow users of BTCPay an easy way to access their wallet through Bitcoin Core RPC. But if I do so, the URL will look like this:

`https://mainnet.demo.btcpayserver.org/rpc` 

By doing so we can use reverse proxy to route the request to the middleware (or even directly to core), without having to ask user to open new port.

This mean that rather than two field, hostname and port, I wished to only have one field URL.

**Comment by craigraw** [comment]
> As you can see here, while the field say URL, an URL doesn't work.

Yes, `Host & Port` would be more accurate, but also arguably more technical language.

> This mean that rather than two field, hostname and port, I wished to only have one field URL.

That's a fairly significant change to the UI, and one I would argue adds complexity for the default use case. The current "host & port" field format is used for configuring Electrum and proxy servers as, so the consistency helps. It may be possible to support this via pasting in a http:// URL, although the `/rpc` path is not easily representable even if it is stored in the configuration.

> I am planning to create some middleware to allow users of BTCPay an easy way to access their wallet through Bitcoin Core RPC.

Do you mean allowing access to Sparrow to create it's own Bitcoin Core wallet in the instance hosted by BTCPay?

**Comment by NicolasDorier** [comment]
@craigraw when a merchant create a wallet on BTCPay, he can choose to add the generated descriptors into Bitcoin Core.
We then expose Bitcoin core RPC over Tor v3, so people can use it with third party wallet such as Sparrow.

The problem is that it only works on Tor, I wished we could make it works on HTTPS as well so it is easier and more reliable to use.

The only alternative, if you don't support HTTP path would be for user to setup a separate domain for Bitcoin RPC, which worse UX than what we have with Tor currently.

**Comment by craigraw** [comment]
> We then expose Bitcoin core RPC over Tor v3

I'm actually not a fan of exposing Bitcoin Core RPC over any untrusted network - if one discovers the host name all wallet privacy is at risk, and an attacker can take control over the node itself. The authentication provided is not particularly well defended. That said I'm willing to support it so long as it doesn't make the default, recommended approach a poorer UX.

A better engineering solution is for BTCPay to host a minimal Electrum server backed by its existing descriptor wallet in Bitcoin Core (ala EPS/BWT, but for descriptor wallets). I've just been writing one myself, see recent commits in the cormorant package. This removes the privacy risk and solves the path issue neatly, while exposing the functionality to a broader range of wallets and avoiding the need to rescan a (potentially pruned) node.

**Comment by NicolasDorier** [comment]
@craigraw actually the authentication would be handled by a middleware I will make, rather than routing the queries directly to Bitcoin Core.

Problem with Electrum server is that it's not HTTP protocol, and so we can't really use reverse proxy, and need users to open ports, which is too difficult for many of them.
On top of it, current electrum solutions often only offer either single wallet support, and/or doesn't work on pruned node.

**Comment by craigraw** [comment]
> need users to open ports, which is too difficult for many of them

Ok understood. An onion address is one way around this, but with the usual Tor issues of course. That said, Electrum protocol over Tor is manageable, Bitcoin Core RPC over Tor is generally not, since the former uses push notifications and the latter polling.

> On top of it, current electrum solutions often only offer either single wallet support, and/or doesn't work on pruned node.

Yes, EPS is single wallet AFAIK, but both EPS and BWT work on a pruned node. That said I wasn't suggesting using either of them, since I understand the intent is to use a Bitcoin Core descriptor wallet (both use legacy wallets). Rather, a new minimal Electrum server backed by the Bitcoin Core descriptor wallet would be a nice addition to the BTCPay offering (IMO).

Wrt to using the RPC interface over HTTP, making use of the descriptor wallet BTCPay creates directly would be too tight an integration IMO - Sparrow would need to create its own Bitcoin Core wallet to be a generic backend for all loaded wallets.

**Comment by NicolasDorier** [comment]
```
Wrt to using the RPC interface over HTTP, making use of the descriptor wallet BTCPay creates directly would be too tight an integration IMO - Sparrow would need to create its own Bitcoin Core wallet to be a generic backend for all loaded wallets.
```
You misunderstood what I meant I think. I am not suggesting that sparrow use BTCPay's API to access the wallet.
What I am saying is that BTCPay already input all the descriptors into Bitcoin core wallet for people who wants.
Sparrow would just use normal RPC integration as it already does.

**Comment by craigraw** [comment]
I think we are on the same page here - I am not talking about using BTCPay's API. But what I am saying is that when using a Bitcoin Core backend, Sparrow actively manages the descriptors in the Bitcoin Core wallet it uses. It is not just a passive listener. It also loads and unloads the Bitcoin Core wallet as necessary. I believe this is true of every wallet application using Bitcoin Core as a wallet tracking backend, but I could be wrong.

Removing this active wallet management means creating too strong a dependance on BTCPay's implementation to guarantee a good UX IMO.

**Comment by NicolasDorier** [comment]
Unloading the wallet BTCPay is importing descriptor to would indeed be a problem. Any other operation should be fine.

I believe that sparrow can offer a better way to spend the bitcoin than our internal wallet in some situation for example.
But the fact that BTCPay import descriptors while Sparrow is connected may raise some issue I agree. (since I guess then you have some state in the UI that should change but there is no notification possible from RPC)
